File: api/clinical_risk.py
import numpy as np
from typing import Dict, List, Tuple


FEATURE_NAMES = ['age', 'gender', 'height', 'weight', 'ap_hi', 'ap_lo',
                 'cholesterol', 'gluc', 'smoke', 'alco', 'active']


def calculate_clinical_risk(patient: Dict) -> Tuple[float, Dict[str, float]]:
    """Calculate 10-year CVD risk with proper clinical weighting.
    
    Returns:
        Tuple of (risk_score, feature_contributions)
    """
    age_years = patient['age'] / 365.25 if patient['age'] > 100 else patient['age']
    systolic_bp = patient['ap_hi']
    diastolic_bp = patient['ap_lo']
    cholesterol = patient['cholesterol']
    glucose = patient['gluc']
    bmi = patient['weight'] / ((patient['height'] / 100) ** 2)
    smoker = patient['smoke']
    alco = patient['alco']
    active = patient['active']
    gender = patient['gender']

    points = 0
    feature_contributions = {
        'age': 0, 'ap_hi': 0, 'ap_lo': 0, 'cholesterol': 0,
        'gluc': 0, 'weight': 0, 'smoke': 0, 'alco': 0, 'active': 0, 'gender': 0
    }

    # Age points - more aggressive for older patients
    if gender == 1:  # Male
        if age_years >= 65:
            points += 10
            feature_contributions['age'] = 10
        elif age_years >= 60:
            points += 8
            feature_contributions['age'] = 8
        elif age_years >= 55:
            points += 6
            feature_contributions['age'] = 6
        elif age_years >= 50:
            points += 4
            feature_contributions['age'] = 4
        elif age_years >= 45:
            points += 2
            feature_contributions['age'] = 2
        elif age_years >= 40:
            points += 1
            feature_contributions['age'] = 1
    else:  # Female
        if age_years >= 65:
            points += 8
            feature_contributions['age'] = 8
        elif age_years >= 60:
            points += 6
            feature_contributions['age'] = 6
        elif age_years >= 55:
            points += 4
            feature_contributions['age'] = 4
        elif age_years >= 50:
            points += 2
            feature_contributions['age'] = 2
        elif age_years >= 45:
            points += 1
            feature_contributions['age'] = 1

    # Blood pressure points - VERY aggressive for hypertensive crisis
    if systolic_bp >= 180:
        points += 10
        feature_contributions['ap_hi'] = 10  # Hypertensive crisis
    elif systolic_bp >= 160:
        points += 7
        feature_contributions['ap_hi'] = 7   # Stage 2 hypertension
    elif systolic_bp >= 140:
        points += 4
        feature_contributions['ap_hi'] = 4   # Stage 1 hypertension
    elif systolic_bp >= 130:
        points += 2
        feature_contributions['ap_hi'] = 2   # Elevated
    elif systolic_bp < 120:
        points -= 1
        feature_contributions['ap_hi'] = -1  # Normal (protective)

    # Diastolic BP penalty
    if diastolic_bp >= 120:
        points += 5
        feature_contributions['ap_lo'] = 5   # Hypertensive crisis
    elif diastolic_bp >= 100:
        points += 3
        feature_contributions['ap_lo'] = 3

    # Cholesterol - well above normal is very high risk
    if cholesterol == 3:
        points += 5
        feature_contributions['cholesterol'] = 5   # Well above normal
    elif cholesterol == 2:
        points += 2
        feature_contributions['cholesterol'] = 2   # Above normal

    # Glucose/diabetes risk
    if glucose == 3:
        points += 4
        feature_contributions['gluc'] = 4   # Diabetic
    elif glucose == 2:
        points += 2
        feature_contributions['gluc'] = 2   # Pre-diabetic

    # BMI - extreme obesity is a major risk factor
    if bmi >= 40:
        points += 6
        feature_contributions['weight'] = 6   # Extreme obesity (class III)
    elif bmi >= 35:
        points += 4
        feature_contributions['weight'] = 4   # Class II
    elif bmi >= 30:
        points += 3
        feature_contributions['weight'] = 3   # Class I
    elif bmi >= 25:
        points += 1
        feature_contributions['weight'] = 1   # Overweight

    # Smoking - MAJOR risk factor
    if smoker:
        points += 5
        feature_contributions['smoke'] = 5
    else:
        feature_contributions['smoke'] = 0

    # Alcohol
    if alco:
        points += 2
        feature_contributions['alco'] = 2
    else:
        feature_contributions['alco'] = 0

    # Physical activity (protective)
    if not active:
        points += 2
        feature_contributions['active'] = 2
    else:
        feature_contributions['active'] = -1  # protective

    # Gender (protective for females)
    if gender == 2:  # Female
        feature_contributions['gender'] = -1
    else:
        feature_contributions['gender'] = 0

    # Convert points to risk probability
    risk = 0.05 + (points / 35.0) * 0.45
    risk = float(np.clip(risk, 0.02, 0.75))

    return risk, feature_contributions


def compute_shap_importance(feature_contributions: Dict[str, float], risk_score: float) -> List[Dict]:
    """Convert feature contributions to SHAP-like values that sum to risk deviation from baseline.
    
    SHAP values represent the actual contribution direction:
    - Positive value = increases risk (red)
    - Negative value = decreases risk (blue/protective)
    They sum to risk_deviation from baseline.
    """
    baseline_risk = 0.15  # approximate baseline
    risk_deviation = risk_score - baseline_risk

    total_abs_contrib = sum(abs(v) for v in feature_contributions.values())
    if total_abs_contrib == 0:
        return [{'feature': k, 'value': 0.0} for k in feature_contributions.keys()]

    # Scale contributions proportionally to match risk_deviation
    # Preserve the SIGN of each contribution (positive = risk increase, negative = protective)
    shap_values = []
    for feature, contrib in feature_contributions.items():
        if total_abs_contrib > 0:
            # Scale each contribution proportionally, preserving its sign
            shap_val = (contrib / total_abs_contrib) * risk_deviation
        else:
            shap_val = 0.0
        shap_values.append({'feature': feature, 'value': float(shap_val)})

    # Sort by absolute value descending, but include all features with non-zero contribution
    shap_values.sort(key=lambda x: abs(x['value']), reverse=True)
    
    # Return top features, but ensure we include at least some protective factors
    # Filter out exactly-zero values only
    non_zero = [s for s in shap_values if abs(s['value']) > 1e-10]
    return non_zero[:10]  # Return up to 10 features


def compute_all_risk_metrics(risk_10yr: float) -> Dict[str, float]:
    """Compute risk metrics for different time horizons."""
    risk_baseline = risk_10yr * 0.15
    risk_2yr = risk_10yr * 0.35
    risk_5yr = risk_10yr * 0.70

    risk_baseline = float(np.clip(risk_baseline, 0.01, 0.50))
    risk_2yr = float(np.clip(risk_2yr, 0.05, 0.60))
    risk_5yr = float(np.clip(risk_5yr, 0.10, 0.70))
    risk_10yr = float(np.clip(risk_10yr, 0.15, 0.75))

    return {
        'immediateRisk': risk_baseline,
        'risk2Year': risk_2yr,
        'risk5Year': risk_5yr,
        'risk10Year': risk_10yr
    }

File: clinical-dashboard/app/api/risk/clinicalRisk.ts
interface PatientData {
  age: number;
  gender: number;
  height: number;
  weight: number;
  ap_hi: number;
  ap_lo: number;
  cholesterol: number;
  gluc: number;
  smoke: number;
  alco: number;
  active: number;
}

interface ShapFeature {
  feature: string;
  value: number;
}

function calculateClinicalRisk(patient: PatientData): [number, Record<string, number>] {
  const ageYears = patient.age > 100 ? patient.age / 365.25 : patient.age;
  const systolicBp = patient.ap_hi;
  const diastolicBp = patient.ap_lo;
  const cholesterol = patient.cholesterol;
  const glucose = patient.gluc;
  const bmi = patient.weight / Math.pow(patient.height / 100, 2);
  const smoker = patient.smoke;
  const alco = patient.alco;
  const active = patient.active;
  const gender = patient.gender;

  let points = 0;
  const featureContributions: Record<string, number> = {
    age: 0, ap_hi: 0, ap_lo: 0, cholesterol: 0,
    gluc: 0, weight: 0, smoke: 0, alco: 0, active: 0, gender: 0
  };

  if (gender === 1) {
    if (ageYears >= 65) { points += 10; featureContributions.age = 10; }
    else if (ageYears >= 60) { points += 8; featureContributions.age = 8; }
    else if (ageYears >= 55) { points += 6; featureContributions.age = 6; }
    else if (ageYears >= 50) { points += 4; featureContributions.age = 4; }
    else if (ageYears >= 45) { points += 2; featureContributions.age = 2; }
    else if (ageYears >= 40) { points += 1; featureContributions.age = 1; }
  } else {
    if (ageYears >= 65) { points += 8; featureContributions.age = 8; }
    else if (ageYears >= 60) { points += 6; featureContributions.age = 6; }
    else if (ageYears >= 55) { points += 4; featureContributions.age = 4; }
    else if (ageYears >= 50) { points += 2; featureContributions.age = 2; }
    else if (ageYears >= 45) { points += 1; featureContributions.age = 1; }
  }

  if (systolicBp >= 180) { points += 10; featureContributions.ap_hi = 10; }
  else if (systolicBp >= 160) { points += 7; featureContributions.ap_hi = 7; }
  else if (systolicBp >= 140) { points += 4; featureContributions.ap_hi = 4; }
  else if (systolicBp >= 130) { points += 2; featureContributions.ap_hi = 2; }
  else if (systolicBp < 120) { points -= 1; featureContributions.ap_hi = -1; }

  if (diastolicBp >= 120) { points += 5; featureContributions.ap_lo = 5; }
  else if (diastolicBp >= 100) { points += 3; featureContributions.ap_lo = 3; }

  if (cholesterol === 3) { points += 5; featureContributions.cholesterol = 5; }
  else if (cholesterol === 2) { points += 2; featureContributions.cholesterol = 2; }

  if (glucose === 3) { points += 4; featureContributions.gluc = 4; }
  else if (glucose === 2) { points += 2; featureContributions.gluc = 2; }

  if (bmi >= 40) { points += 6; featureContributions.weight = 6; }
  else if (bmi >= 35) { points += 4; featureContributions.weight = 4; }
  else if (bmi >= 30) { points += 3; featureContributions.weight = 3; }
  else if (bmi >= 25) { points += 1; featureContributions.weight = 1; }

  if (smoker) { points += 5; featureContributions.smoke = 5; }
  if (alco) { points += 2; featureContributions.alco = 2; }

  if (active === 1) { points -= 2; featureContributions.active = -2; }
  else { points += 2; featureContributions.active = 2; }

  if (gender === 2) featureContributions.gender = -1;

  let risk = 0.05 + (points / 35.0) * 0.45;
  risk = Math.max(0.02, Math.min(0.75, risk));

  return [risk, featureContributions];
}

function computeShapImportance(featureContributions: Record<string, number>, riskScore: number): ShapFeature[] {
  const totalAbsContrib = Object.values(featureContributions).reduce((sum, v) => sum + Math.abs(v), 0);
  if (totalAbsContrib === 0) {
    return Object.keys(featureContributions).map(k => ({ feature: k, value: 0.0 }));
  }

  const shapValues: ShapFeature[] = [];
  for (const [feature, contrib] of Object.entries(featureContributions)) {
    const shapVal = (contrib / totalAbsContrib) * riskScore;
    shapValues.push({ feature, value: shapVal });
  }

  shapValues.sort((a, b) => Math.abs(b.value) - Math.abs(a.value));
  return shapValues.filter(s => Math.abs(s.value) > 1e-10).slice(0, 10);
}

function computeAllRiskMetrics(risk10yr: number): Record<string, number> {
  const riskBaseline = risk10yr * 0.15;
  const risk2yr = risk10yr * 0.35;
  const risk5yr = risk10yr * 0.70;

  return {
    immediateRisk: Math.max(0.01, Math.min(0.50, riskBaseline)),
    risk2Year: Math.max(0.05, Math.min(0.60, risk2yr)),
    risk5Year: Math.max(0.10, Math.min(0.70, risk5yr)),
    risk10Year: Math.max(0.15, Math.min(0.75, risk10yr))
  };
}

export function calculateRisk(patient: PatientData) {
  const [risk10yr, featureContributions] = calculateClinicalRisk(patient);
  const riskMetrics = computeAllRiskMetrics(risk10yr);
  const shapImportance = computeShapImportance(featureContributions, risk10yr);

  return {
    ...riskMetrics,
    shapImportance
  };
}